src/Npgsql/BackendMessages/RowDescriptionMessage.cs

```diff
@@ -301,38 +297,67 @@ internal FieldDescription(FieldDescription source)
     /// </summary>
     internal FormatCode FormatCode { get; set; }
 
-    internal string TypeDisplayName => PostgresType.GetDisplayNameWithFacets(TypeModifier);
+    internal DataFormat Format => FormatCode is FormatCode.Binary ? DataFormat.Binary : DataFormat.Text;
 
-    /// <summary>
-    /// The Npgsql type handler assigned to handle this field.
-    /// Returns <see cref="UnknownTypeHandler"/> for fields with format text.
-    /// </summary>
-    internal NpgsqlTypeHandler Handler { get; private set; }
+    internal Field Field => new(Name, _serializerOptions.GetCanonicalTypeId(PostgresType), TypeModifier);
 
+    internal string TypeDisplayName => PostgresType.GetDisplayNameWithFacets(TypeModifier);
+
+    PostgresType? _postgresType;
     internal PostgresType PostgresType
-        => _typeMapper.DatabaseInfo.ByOID.TryGetValue(TypeOID, out var postgresType)
-            ? postgresType
-            : UnknownBackendType.Instance;
+        => _postgresType ??= _serializerOptions.TryGetPgType((Oid)TypeOID) ?? UnknownBackendType.Instance;
 
-    internal Type FieldType => Handler.GetFieldType(this);
+    internal Type FieldType => ObjectOrDefaultTypeInfo.Type;
 
-    internal void ResolveHandler()
-        => Handler = IsBinaryFormat ? _typeMapper.ResolveByOID(TypeOID) : _typeMapper.UnrecognizedTypeHandler;
+    PgTypeInfo? _objectOrDefaultTypeInfo;
+    PgTypeInfo ObjectOrDefaultTypeInfo => _objectOrDefaultTypeInfo ??= GetObjectOrDefaultConverterInfo(_serializerOptions, PostgresType);
 
-    TypeMapper _typeMapper;
+    PgConverterInfo? _objectOrDefaultInfo;
+    internal PgConverterInfo ObjectOrDefaultInfo => _objectOrDefaultInfo ??=
+        // Always go through object apis.
+        ObjectOrDefaultTypeInfo.Bind(Field, Format) with { AsObject = true };
 
-    internal bool IsBinaryFormat => FormatCode == FormatCode.Binary;
-    internal bool IsTextFormat => FormatCode == FormatCode.Text;
+    PgSerializerOptions _serializerOptions;
+    PgTypeInfo? _lastTypeInfo;
+    PgConverterInfo _lastConverterInfo;
 
     internal FieldDescription Clone()
     {
-        var field =  new FieldDescription(this);
-        field.ResolveHandler();
+        var field = new FieldDescription(this);
         return field;
     }
 
+    internal PgConverterInfo GetOrAddConverterInfo(Type type)
```

Comment:
I used to have GetConverterInfo as well, it might be useful to use in TextReader/GetBytes() places, to not stomp the cache for uncommon operations. Not sure yet, if it doesn't seem useful we can definitely rename.